web/src/recent_view_data.ts

```diff
@@ -1,16 +1,41 @@
-import type {Message} from "./message_store";
+import assert from "minimalistic-assert";
+
+import type {DisplayRecipientUser, Message} from "./message_store";
 import * as people from "./people";
-import {get_key_from_message} from "./recent_view_util";
+import {get_key_from_conversation_data} from "./recent_view_util";
+import * as unread from "./unread";
 
 export type ConversationData = {
-    last_msg_id: number;
     participated: boolean;
-    type: "private" | "stream";
-};
+    unread_count: number;
+    latest_mesage_timestamp: number;
+    last_msg_id: number; // should not be used for fetching messages
+} & (
+    | {
+          type: "private";
+          to_user_ids: string;
+          display_recipient: DisplayRecipientUser[];
+          display_reply_to: string;
+          recipient_id: number;
+          pm_with_url: string;
+      }
+    | {
+          type: "stream";
+          stream_id: number;
+          topic: string;
+      }
+);
 export const conversations = new Map<string, ConversationData>();
 // For stream messages, key is stream-id:topic.
 // For pms, key is the user IDs to whom the message is being sent.
 
+function message_to_conversation_unread_count(msg: Message): number {
```

Comment:
This was moved from `web/src/recent_view_ui.ts` since it's used in this file now.